src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BooleanClassifier.cs

```diff
@@ -1,62 +1,52 @@
 {
-    /// <summary>Classifies characters into true or false</summary>
+    /// <summary>Classifies characters as true or false.</summary>
+    /// <remarks>
+    /// The classification is determined entirely by the BDD used to construct the classifier, and in fact
+    /// simply calling Contains on the BDD instead of using the classifier would suffice from a correctness
+    /// perspective. The classifier as a wrapper for the BDD is valuable in order to optimize for ASCII, as
+    /// it precomputes the results for ASCII inputs and stores them in a separate table, only falling back
+    /// to using the BDD for non-ASCII.
+    /// </remarks>
     internal sealed class BooleanClassifier
     {
-        /// <summary>Explicit array for ascii.</summary>
+        /// <summary>Lookup table used for ASCII characters.</summary>
         private readonly bool[] _ascii;
-        /// <summary>Stores the remaining characters in a BDD.</summary>
-        private readonly BDD _nonAsciiBDD;
-
-        private BooleanClassifier(ulong lower, ulong upper, BDD bdd)
-        {
-            var ascii = new bool[128];
-            for (int i = 0; i < ascii.Length; i++)
-            {
-                ascii[i] = (i < 64 ? (lower & ((ulong)1 << i)) : (upper & ((ulong)1 << (i - 64)))) != 0;
-            }
-
-            _ascii = ascii;
-            _nonAsciiBDD = bdd;
-        }
-
-        private BooleanClassifier(bool[] ascii, BDD bdd)
-        {
-            Debug.Assert(ascii.Length == 128);
-            _ascii = ascii;
-            _nonAsciiBDD = bdd;
-        }
+        /// <summary>BDD used for non-ASCII characters.</summary>
+        private readonly BDD _nonAscii;
 
         /// <summary>Create a Boolean classifier.</summary>
-        /// <param name="solver">character algebra (the algebra is not stored in the classifier)</param>
-        /// <param name="domain">elements that map to true</param>
-        internal static BooleanClassifier Create(CharSetSolver solver, BDD domain)
+        /// <param name="solver">Character algebra (the algebra is not stored in the classifier)</param>
+        /// <param name="domain">Elements that map to true.</param>
+        public BooleanClassifier(CharSetSolver solver, BDD domain)
         {
+            // We want to optimize for ASCII, so query the BDD for each ASCII character in
+            // order to precompute a lookup table we'll use at match time.
             var ascii = new bool[128];
             for (int i = 0; i < ascii.Length; i++)
             {
                 ascii[i] = domain.Contains(i);
             }
 
-            // Remove the ASCII characters from the domain if the domain is not everything
-            BDD bdd = domain.IsFull ?
-                domain :
-                solver.And(solver._nonascii, domain);
+            // Now, as an optimization, we can remove the ASCII characters from the BDD
+            // as they'll never be used.
+            domain = solver.And(solver._nonAscii, domain);
```

Comment:
trying to follow.. why remove IsFull check here?

also, should we be renaming domain (to bdd?)

Reply:
Margus and Olli can correct me if I'm wrong, but I believe it's superfluous.  And immediately calls a helper to apply the And operation, and that helper immediately checks to see if either input is equal to True, returning the other if it is.  IsFull is the same as checking for equality to True.  It seemed strange then to just special-case this one condition.

As for renaming domain to bdd, I considered it and didn't end up changing it; happy to change it to bdd or something else if folks prefer.

Reply:
Oh, Dan is right here, I missed this edit, ...good to have multiple people look at stuff. 
We should keep the Full case because it is the trivial True BDD instead of replacing it with the more complicated one here. Usually, unless domain is just True, the replacement after cutting of the ASCII portion gives a smaller BDD, but not when it is just True, the the domai just becomes nonASCII, but we already know it is nonASCII so it would check twice

Reply:
I'm not following... 

```C#
solver.And(solver._nonAscii, domain)
```
just ends up being a call to:
https://github.com/dotnet/runtimelab/blob/982544a8d226c4edc74fdd38e91778c0d496978f/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs#L74
```C#
ApplyBinaryOp(Op.And, solver._nonAscii, domain)
```
which in turn does:
https://github.com/dotnet/runtimelab/blob/982544a8d226c4edc74fdd38e91778c0d496978f/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs#L94-L120
```C#
                    if (b == True)
                        return a;
```
meaning "if domain == True return _nonAscii".  So the optimization is still there, no?

Reply:
Stephen, regarding 
>It seemed strange then to just special-case this one condition.
note: the operation there was special casing conjunction (indeed that would not make much sense)
but rather, if the domain == True then keep it "as is" , do not cut it to nonASCII

Reply:
Thanks, I understand.  My point was, why would the call site only special case IsFull and not every other special case that's in the case BoolOp.And section of ApplyBinaryOp... we can just always call And/ApplyBinaryOp and let any or all of its relevant optimizations kick in.

Reply:
> but here we reall just want to keep TRue

I understand. And with my edit we still do.  Am I going crazy?